nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -1872,63 +2111,132 @@ impl<'a> Planner<'a> {
+        let new_repo = self.input.tuf_repo().description();
+        for sled_id in self.blueprint.sled_ids_with_zones() {
+            for z in self.blueprint.current_sled_zones(
+                sled_id,
+                BlueprintZoneDisposition::is_in_service,
+            ) {
+                let kind = z.zone_type.kind();
+                if kind != ZoneKind::Nexus
+                    && z.image_source != new_repo.zone_image_source(kind)?
+                {
                     return Ok(false);
                 }
+            }
+        }
+        return Ok(true);
+    }
 
-                // Check that all in-service zones (other than Nexus) on all
-                // sleds have an image source consistent with `new_repo`.
-                for sled_id in self.blueprint.sled_ids_with_zones() {
-                    for z in self.blueprint.current_sled_zones(
-                        sled_id,
-                        BlueprintZoneDisposition::is_in_service,
-                    ) {
-                        let kind = z.zone_type.kind();
-                        if kind != ZoneKind::Nexus
-                            && z.image_source
-                                != new_repo.zone_image_source(kind)?
-                        {
-                            return Ok(false);
-                        }
-                    }
+    fn lookup_current_nexus_image(&self) -> Option<BlueprintZoneImageSource> {
+        // Look up the active Nexus zone in the blueprint to get its image
+        self.blueprint
+            .parent_blueprint()
+            .all_omicron_zones(BlueprintZoneDisposition::is_in_service)
+            .find_map(|(_, blueprint_zone)| {
+                if self.input.active_nexus_zones().contains(&blueprint_zone.id)
+                {
+                    Some(blueprint_zone.image_source.clone())
+                } else {
+                    None
                 }
+            })
+    }
+
+    fn lookup_current_nexus_generation(
+        &self,
+    ) -> Result<Option<Generation>, Error> {
+        // Look up the active Nexus zone in the blueprint to get its generation
+        self.blueprint
+            .parent_blueprint()
+            .find_generation_for_nexus(self.input.active_nexus_zones())
+            .map_err(|_| Error::NoActiveNexusZonesInParentBlueprint)
+    }
 
-                Ok(true)
+    // Returns if the zone is ready to be updated.
+    //
+    // If the zone should not be updated yet, updates the planner report to
+    // identify why it is not ready for update.
+    fn can_zone_be_updated(
+        &self,
+        zone: &BlueprintZoneConfig,
+        report: &mut PlanningZoneUpdatesStepReport,
+    ) -> Result<bool, Error> {
+        let safe_to_update = self.can_zone_be_shut_down_safely(zone, report);
+
+        let zone_nexus_generation = match &zone.zone_type {
+            // For Nexus, we need to confirm that the active generation has
+            // moved beyond this zone.
```

Comment:
Calling this function for Nexus feels a little weird.  For other zones, I think we're saying: "we're about to update this one zone, either by expunging it [and then subsequently adding one] or else replacing it.  Can we do that now?"  For Nexus, though, this will only return `true` *after* we've done the handoff to the new fleet.

I'm not sure it's worth reworking.  Assuming not, I'd clarify this comment a bit:

```suggestion
            // For Nexus, we're only ready to "update" this zone once control has been handed off to a newer generation of Nexus zones.  (Once that happens, we're not really going to update this zone, just expunge it.)
```

Reply:
Refactored this a bit, but done in 4beb4209.